### PR TITLE
Compare documents

### LINKED ISSUE #1521: Implement document comparison
Currently, `types.Compare` skips document comparison. It should implement it. See https://www.mongodb.com/docs/manual/reference/bson-type-comparison-order/#objects and the whole page in general.

Cover with compatibility tests, check test coverage and simplify existing code as you go.

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

Closes #1521.

## Readiness checklist

* [ ] I added tests for new functionality or bugfixes.
* [x] I ran `task all`, and it passed.
* [x] I added/updated comments for both exported and unexported top-level declarations (functions, types, etc).
* [ ] I checked comments rendering with `task godocs`.
* [x] I ensured that the title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [ ] I marked all done items in this checklist.

### CODE REVIEW

File: internal/types/compare.go [hunk 1]
```diff
@@ -55,9 +55,11 @@ func Compare(docValue, filterValue any) CompareResult {
 
 	switch docValue := docValue.(type) {
 	case *Document:
-		// TODO: implement document comparing https://github.com/FerretDB/FerretDB/issues/457
-		return Incomparable
+		if filterDoc, ok := filterValue.(*Document); ok {
+			return compareDocuments(docValue, filterDoc)
+		}
 
+		return compareTypeOrder(docValue, filterValue)
```
Comment (chilagrow) [inline internal/types/compare.go:62]: Maybe this line should be incomparable.

File: integration/query_comparison_compat_test.go [hunk 2]
```diff
@@ -491,8 +527,26 @@ func TestQueryComparisonCompatLt(t *testing.T) {
 
 	t.Parallel()
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/1521
 	testCases := map[string]queryCompatTestCase{
+		"Document": {
+			filter: bson.D{{"v", bson.D{{"$eq", bson.D{{"foo", int32(42)}, {"42", "foo"}, {"array", bson.A{int32(42), "foo", nil}}}}}}},
```
Comment (w84thesun) [inline integration/query_comparison_compat_test.go:532]: Should we use "$lt" here?

File: integration/query_comparison_compat_test.go [hunk 3]
```diff
@@ -393,8 +411,26 @@ func TestQueryComparisonCompatGte(t *testing.T) {
...
```
Comment (w84thesun) [inline integration/query_comparison_compat_test.go:416]: Should we use "$gte" here?

File: integration/query_comparison_compat_test.go [hunk 4]
```diff
@@ -597,8 +651,26 @@ func TestQueryComparisonCompatLte(t *testing.T) {
...
```
Comment (w84thesun) [inline integration/query_comparison_compat_test.go:656]: Should we use "$lte" here?

**Review by w84thesun — COMMENTED** [review]
Added few questions about test cases operators.

File: integration/query_comparison_compat_test.go (hunk 4 above)
Comment (chilagrow) [reply integration/query_comparison_compat_test.go:656]: No wonder tests didn't fail :see_no_evil: thanks for catching this

**Comment by chilagrow** [comment]
Need more work on this.

**Review by rumyantseva — COMMENTED** [review]
The code looks good to me. There is a comment from Dmitry and a potential merge conflict with the second task, but the implementation looks correct to me.

File: internal/handlers/common/filter.go [hunk 5]
```diff
@@ -348,32 +352,38 @@ func filterFieldExpr(doc *types.Document, filterKey string, expr *types.Document
...
-					arr := arrValue.FilterArrayByType(exprValue)
-
-					if arr.Len() == 0 {
-						// The array does not contain any element with the same type as exprValue.
-						return false, nil
-					}
+			}
 
-					fieldValue = arr.Max()
+			if ok := filterByComparisonOperandApplicable(fieldValue, exprValue); !ok {
```
Comment (chilagrow) [inline internal/handlers/common/filter.go:360]: See https://www.mongodb.com/docs/manual/reference/method/db.collection.find/#type-bracketing
> The document with _id equal to "avocados" is not returned because its qty value is of type string while the [$gt](https://www.mongodb.com/docs/manual/reference/operator/query/gt/#mongodb-query-op.-gt) operand is of type integer.

File: internal/handlers/common/filter.go [hunk 6]
```diff
@@ -348,32 +352,38 @@ func filterFieldExpr(doc *types.Document, filterKey string, expr *types.Document
...
+			arrField, isFieldArray := fieldValue.(*types.Array)
+
+			if !isExprArray && isFieldArray {
+				// exprValue is not an array but fieldValue is an array
+				arr := arrField.FilterArrayByType(exprValue)
+				if arr.Len() == 0 {
+					return false, nil
 				}
+
+				fieldValue = arr.Max()
```
Comment (chilagrow) [inline internal/handlers/common/filter.go:386]: See https://www.mongodb.com/docs/manual/reference/bson-type-comparison-order/#arrays
>A greater-than comparison, or a descending sort, compares the largest elements of the array according to the reverse BSON type sort order.

File: internal/handlers/common/filter.go [hunk 7]
```diff
@@ -348,32 +352,38 @@ func filterFieldExpr(doc *types.Document, filterKey string, expr *types.Document
...
+			// and results in Less. Other values "foo" and nil which are
+			// not number type are not considered for $gt comparison.
+			_, isExprArray := exprValue.(*types.Array)
+			arrField, isFieldArray := fieldValue.(*types.Array)
+
+			if !isExprArray && isFieldArray {
+				// exprValue is not an array but fieldValue is an array
+				arr := arrField.FilterArrayByType(exprValue)
+				if arr.Len() == 0 {
+					return false, nil
```
Comment (chilagrow) [inline internal/handlers/common/filter.go:383]: See https://www.mongodb.com/docs/manual/reference/bson-type-comparison-order/#arrays

> A comparison of an empty array (example, [ ]) considers the empty array as less than a null value or a missing field value.

File: integration/shareddata/composites.go [hunk 8]
```diff
@@ -42,7 +43,43 @@ var Composites = &Values[string]{
 		"array-null":          bson.A{nil},
 		"array-numbers-asc":   bson.A{int32(42), int64(43), 45.5},
 		"array-strings-desc":  bson.A{"c", "b", "a"},
-		"array-composite":     bson.A{nil, primitive.Timestamp{T: 42, I: 13}, false, "z", primitive.NilObjectID},
+		"array-documents":     bson.A{bson.D{{"field", int32(42)}}, bson.D{{"field", int32(44)}}},
+		"array-composite": bson.A{
```
Comment (rumyantseva) [inline integration/shareddata/composites.go:47]: As agreed in the call, maybe this document could be shorter and could contain only one value of each type + corner cases (so it's easier to maintain it in the future)

File: internal/handlers/common/filter.go [hunk 9]
```diff
@@ -1435,3 +1467,53 @@ func filterFieldExprElemMatch(doc *types.Document, filterKey string, exprValue a
...
+		_, ok := fieldValue.(types.NullType)
+		return ok
+	case types.Timestamp:
+		_, ok := fieldValue.(types.Timestamp)
+		return ok
+	default:
+		panic("unsupported type")
+	}
+
+	return true
```
Comment (rumyantseva) [inline internal/handlers/common/filter.go:1518]: Let's remove this return as it's hard to follow from which case it's coming. Every case has their own return and then this one is not needed.

**Review by rumyantseva — CHANGES_REQUESTED** [review]
Great job overall! Only few questions we discussed in the call.

File: internal/handlers/common/filter.go [hunk 10]
```diff
@@ -1435,3 +1467,53 @@ func filterFieldExprElemMatch(doc *types.Document, filterKey string, exprValue a
...
+// filterByComparisonOperandApplicable returns true if comparison operands
+// $gt, $gte, $lt and $lte can be applied to fieldValue and exprValue.
+func filterByComparisonOperandApplicable(fieldValue, exprValue any) bool {
+	if _, ok := fieldValue.(*types.Array); ok {
+		// array is special since it could contain values
+		// which maybe the same type as operand.
+		return true
+	}
+
+	switch exprValue.(type) {
```
Comment (rumyantseva) [inline internal/handlers/common/filter.go:1480]: Also as discussed in the call, it looks like the linter doesn't complain about types order in this case, so we need to create a new issue to investigate why the linter doesn't work.

File: internal/handlers/common/filter.go [hunk 11]
```diff
@@ -1435,3 +1467,53 @@ func filterFieldExprElemMatch(doc *types.Document, filterKey string, exprValue a
 
 	return filterFieldExpr(doc, filterKey, expr)
 }
+
+// filterByComparisonOperandApplicable returns true if comparison operands
+// $gt, $gte, $lt and $lte can be applied to fieldValue and exprValue.
+func filterByComparisonOperandApplicable(fieldValue, exprValue any) bool {
```
Comment (rumyantseva) [inline internal/handlers/common/filter.go:1473]: I like the idea of that function a lot! It simplifies understanding!

File: internal/types/compare.go [hunk 12]
```diff
@@ -306,3 +295,83 @@ func compareArrays(filterArr, docArr *Array) CompareResult {
...
+
+	if a.Len() < b.Len() {
+		return Less
+	}
+
+	return Equal
+}
+
+// compareArrayToNonArray compares array to a non-array value.
+func compareArrayToNonArray(as *Array, b any) CompareResult {
```
Comment (rumyantseva) [inline internal/types/compare.go:348]: What happens if b is an array too? If the function doesn't support that case, should we panic?

File: internal/handlers/common/filter.go [hunk 13]
```diff
@@ -1435,3 +1467,53 @@ func filterFieldExprElemMatch(doc *types.Document, filterKey string, exprValue a
...
+	if _, ok := fieldValue.(*types.Array); ok {
+		// array is special since it could contain values
+		// which maybe the same type as operand.
+		return true
+	}
+
+	switch exprValue.(type) {
+	case float64, int32, int64:
+		switch fieldValue.(type) {
+		case float64, int32, int64:
```
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/common/filter.go:1489]: Fixed

File: internal/handlers/common/filter.go (hunk 9 above)
Comment (chilagrow) [reply internal/handlers/common/filter.go]: Removed.

File: internal/handlers/common/filter.go (hunk 10 above)
Comment (chilagrow) [reply internal/handlers/common/filter.go:1480]: Fixed the order and created https://github.com/FerretDB/FerretDB/issues/1630

File: internal/types/compare.go (hunk 12 above)
Comment (chilagrow) [reply internal/types/compare.go:348]: Maybe I can call `compareArrays(as,b)`. At the moment it might panic :scream:

Comment (rumyantseva) [reply internal/types/compare.go:348]: I mean, it's fine if we panic in such a case, if we expect the caller to be responsible for `b` not being the array, let's make such a panic explicit

Comment (chilagrow) [reply internal/types/compare.go:348]: Actually since the code can handle comparing two arrays, I think it makes sense for this function to call `compareArrays` to compare them array. No need to panic here.

**Comment by chilagrow** [comment]
Build fails currently :thinking: 

```
#2 [internal] load .dockerignore
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/debian:11.5
#3 ...

#4 [auth] library/debian:pull token for registry-1.docker.io
#4 DONE 0.0s

#3 [internal] load metadata for docker.io/library/debian:11.5
#3 DONE 1.0s

#5 [1/1] FROM docker.io/library/debian:11.5@sha[25](https://github.com/FerretDB/FerretDB/actions/runs/3637778243/jobs/6139182202#step:6:26)6:a[28](https://github.com/FerretDB/FerretDB/actions/runs/3637778243/jobs/6139182202#step:6:29)8aa7ad0e4d443e86843972c25a02f99e9ad6ee589dd764895b2c3f5a8340b
#5 resolve docker.io/library/debian:11.5@sha256:a288aa7ad0e4d443e86843972c25a02f99e9ad6ee589dd764895b2c3f5a8340b done
#5 sha256:a4b14cf13428c71a5fce6d9380eb2ac1452a7ae4253e594dc90c8a884fdf3db2 5[29](https://github.com/FerretDB/FerretDB/actions/runs/3637778243/jobs/6139182202#step:6:30)B / 529B done
#5 sha256:291bf168077c554abde831d52178c76aa8d7dad0e55a2e37a68c2b55a5d92e64 1.46kB / 1.46kB done
#5 sha256:f2f58072e9ed1aa1b0143341c5ee83815c00ce47548[30](https://github.com/FerretDB/FerretDB/actions/runs/3637778243/jobs/6139182202#step:6:31)9fa240155067ab0e698 0B / 55.04MB 0.1s
#5 sha256:a288aa7ad0e4d443e86843972c25a02f99e9ad6ee589dd764895b2c3f5a8340b 1.85kB / 1.85kB done
#5 sha256:f2f58072e9ed1aa1b0143341c5ee83815c00ce47548309fa240155067ab0e698 7.34MB / 55.04MB 0.2s
#5 sha256:f2f58072e9ed1aa1b0143341c5ee83815c00ce47548309fa240155067ab0e698 15.73MB / 55.04MB 0.3s
#5 sha256:f2f58072e9ed1aa1b0143341c5ee83815c00ce47548309fa240155067ab0e698 24.12MB / 55.04MB 0.4s
#5 sha256:f2f58072e9ed1aa1b0143341c5ee83815c00ce47548309fa240155067ab0e698 32.51MB / 55.04MB 0.5s
#5 sha256:f2f58072e9ed1aa1b0143341c5ee83815c00ce47548309fa240155067ab0e698 55.04MB / 55.04MB 0.7s
#5 sha256:f2f58072e9ed1aa1b0143341c5ee83815c00ce47548309fa240155067ab0e698 55.04MB / 55.04MB 0.7s done
#5 extracting sha256:f2f58072e9ed1aa1b0143341c5ee83815c00ce47548309fa240155067ab0e698
#5 extracting sha256:f2f58072e9ed1aa1b0143341c5ee83815c00ce47548309fa240155067ab0e698 1.6s done
#5 DONE 2.4s

#6 exporting to image
#6 exporting layers done
#6 writing image sha256:5a652491d05f00366d8d337346f4a11f70aa635b1[31](https://github.com/FerretDB/FerretDB/actions/runs/3637778243/jobs/6139182202#step:6:32)b7cf9f7a7ab69d0344a05 done
#6 naming to docker.io/library/ferretdb-debian done
#6 DONE 0.0s
Selecting previously unselected package ferretdb.
(Reading database ... 6661 files and directories currently installed.)
Preparing to unpack /deb/ferretdb.deb ...
Unpacking ferretdb (0.7.0~79-g7aafd6f) ...
Setting up ferretdb (0.7.0~79-g7aafd6f) ...
ferretdb: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.[32](https://github.com/FerretDB/FerretDB/actions/runs/3637778243/jobs/6139182202#step:6:33)' not found (required by ferretdb)
ferretdb: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.[34](https://github.com/FerretDB/FerretDB/actions/runs/3637778243/jobs/6139182202#step:6:35)' not found (required by ferretdb)
task: Failed to run task "packages": task: Failed to run task "packages-deb": exit status 1
Error: Process completed with exit code 1.
```

**Review by rumyantseva — DISMISSED** [review]
The logic looks correct to me, and I also like documentation and explanation.

**Comment by chilagrow** [comment]
> Build and codev fail currently thinking
> 
> Packages / Build packages (pull_request) Failing after 3m
> 
> ```
> #5 [1/1] FROM docker.io/library/debian:11.5@sha[25]
> ...
> ferretdb: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.[32](https://github.com/FerretDB/FerretDB/actions/runs/3637778243/jobs/6139182202#step:6:33)' not found (required by ferretdb)
> ferretdb: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.[34](https://github.com/FerretDB/FerretDB/actions/runs/3637778243/jobs/6139182202#step:6:35)' not found (required by ferretdb)
> task: Failed to run task "packages": task: Failed to run task "packages-deb": exit status 1
> Error: Process completed with exit code 1.
> ```
> 
> codev
> 
> ```
> [2022-12-07T10:04:57.851Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 503 - upstream connect error or disconnect/reset before headers. reset reason: connection failure
> [2022-12-07T10:04:57.852Z] ['verbose'] The error stack is: Error: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 503 - upstream connect error or disconnect/reset before headers. reset reason: connection failure
>     at main (/snapshot/repo/dist/src/index.js)
>     at processTicksAndRejections (node:internal/process/task_queues:96:5)
> [2022-12-07T10:04:57.852Z] ['verbose'] End of uploader: 337 milliseconds
> Error: Codecov: Failed to properly upload: The process '/home/runner/work/_actions/codecov/codecov-action/v3/dist/codecov' failed with exit code 255
> ```

Fixed on main now thanks :pray:

**Review by w84thesun — DISMISSED** [review]

**Comment by chilagrow** [comment]
Skipped integration tests failing on tigris.

**Review by w84thesun — APPROVED** [review]
